### ISSUE TITLE
Work: 我发现tracking奖励锚点都是世界坐标系，对sim2real不友好，能否改成相对的

### ISSUE THREAD

**Issue by original-dk** [Motphys/UniLab#607]
### Work type

feature

### Area

G1

### Problem

_reward_motion_global_root_ori
_reward_motion_global_root_pos
...
特别的，obs中有个6元数，也是世界的，我们是否可以相对初始态yaw？这样测试部署更加友好，当然可能会牺牲一些规则，导致tracking效果下降。

### Deliverable

    def _update_diff_base_pos(self):
        # 1. 计算各自在世界坐标系下的相对位移（相对于自己的 init_pos）
        rel_robot_world = self.base_pos - self.init_base_pos  # 当前机器人
        rel_traj_world = self.traj.traj_qpos[:, :3] - self.traj.traj_init_qpos[:, :3]  # 轨迹参考

        # 2. 分别转换到「自己默认朝向」的局部坐标系（即各自的 init_quat 坐标系）
        #    使用 transform_by_quat + inv_init_quat 把世界向量转到局部坐标系
        rel_robot_local = transform_by_quat(rel_robot_world, self.inv_base_init_quat)
        rel_traj_local = transform_by_quat(rel_traj_world, self.traj.traj_inv_base_init_quat)

        # 3. 在各自局部坐标系下求差值 → 这就是你想要的 "各自算自己默认朝向后的 rel，再算 diff"
        diff_local = rel_traj_local - rel_robot_local
        self.extras["episode"]["diff_base_pos"].copy_(diff_local)

    def _update_diff_base_rotation(self):
        # 1. 计算当前的相对旋转和轨迹的相对旋转
        rel_current_quat = _tc_quat_mul(self.inv_base_init_quat, self.base_quat)
        rel_traj_quat = _tc_quat_mul(self.traj.traj_inv_base_init_quat, self.traj.traj_qpos[:, 3:7])

        # 2. 计算误差四元数: q_err = q_current^{-1} * q_target
        q_err = _tc_quat_mul(inv_quat(rel_current_quat), rel_traj_quat)

        # 3. 提取 w 和 xyz 部分 (假设 Genesis 四元数格式为 [w, x, y, z])
        q_w = q_err[:, 0:1]
        q_xyz = q_err[:, 1:4]

        # 4. 确保最短路径: 如果 w < 0，说明转了超 180 度，对 xyz 取反
        sign = torch.sign(q_w)
        sign = torch.where(sign == 0, torch.ones_like(sign), sign)  # 防止 w=0 时变成 0 向量

        # 5. 生成 3D 旋转误差向量 [num_envs, 3]
        diff_3d = 2.0 * sign * q_xyz

        # 赋值给 diff 字典（维度完美匹配你 __init__ 里定义的 [num_envs, 3]）
        self.extras["episode"]["diff_base_rotation"].copy_(diff_3d)     该伪代码由grok生成！

### Definition of done

...

**Comment by LeeLeno** [comment]
@original-dk Thanks for the proposal.

**On "the 6D in obs is in world frame" — this doesn't match the code.**

See `_write_motion_anchor_transform` (`tracking.py:280-330`):

```python
# position: rotated into the robot's current body frame
out_pos  = inv(robot_anchor_quat_w) · (anchor_pos_w − robot_anchor_pos_w)
# orientation 6D: relative rotation from robot's current orientation to the reference
out_ori6 = to6d( inv(robot_anchor_quat_w) · anchor_quat_w )
```

The actor obs (`tracking.py:703-740`) = `command + anchor_pos_b + anchor_ori6_b + local_linvel + gyro + joint + action` — all egocentric, no absolute world-frame quantities. So the deployed policy's inputs are already relative; the sim2real concern on the obs side doesn't really apply here.

**On `_reward_motion_global_root_pos/ori` (`tracking.py:1261-1294`):**

These are indeed world-frame, but they're training-time reward shaping, not policy I/O — they aren't computed at deployment. And since the obs contain no absolute orientation, the policy physically can't condition on world heading; this reward can only be realized through the relative obs, so the resulting behavior is already close to relative tracking. The global anchor is the same design as BeyondMimic: it pins absolute heading and suppresses yaw drift. Switching to a purely "relative-to-each-own-init" formulation removes that constraint and may degrade tracking — which you also noted.

**Bottom line:** decoupling absolute heading is better handled at deploy time by re-anchoring the reference to the robot's current pose, rather than changing the reward. The current design is hardware-validated, deployed, and reported in the paper (including walltime).

If there's real demand later, a safer path would be to expose relative-to-init anchoring as an optional config flag rather than replacing the validated default.

**Comment by original-dk** [comment]
Thanks for your tip,I get it!
**sim2sim or sim2real:**
I tried writing the following code:
```
    def get_q6d(self, data):
        # 1. 机器人当前真实姿态, 提取机器人当前的纯 Yaw
        robot_quat = data.body('torso_link').xquat.astype(np.float32)

        robot_yaw_quat = robot_quat.copy()
        robot_yaw_quat[1:3] = 0.0
        robot_yaw_quat = normalize(robot_yaw_quat)

        # 2. 提取轨迹的纯 Roll/Pitch（去掉轨迹自己的 yaw）
        traj_quat = self.traj.traj_anchor_quat[self.traj.traj_index].astype(np.float32)

        traj_yaw_quat = traj_quat.copy()
        traj_yaw_quat[1:3] = 0.0
        traj_yaw_quat = normalize(traj_yaw_quat)

        traj_yaw_inverse = inv_quat(traj_yaw_quat)

        # 正确顺序：inv(traj_yaw) * traj_quat → 得到轨迹的 tilt 部分
        traj_rp_quat = transform_quat_by_quat(traj_yaw_inverse, traj_quat)

        # 3. 构造 re-anchored 的参考：机器人当前 yaw + 轨迹 tilt
        # 正确顺序：robot_yaw * traj_rp
        hybrid_ref_quat = transform_quat_by_quat(robot_yaw_quat, traj_rp_quat)

        # 4. 计算相对旋转（保持和原来一致的相对形式）
        ori_q = transform_quat_by_quat(
            inv_quat(robot_quat[None, :]),
            hybrid_ref_quat[None, :]
        )

        motion_anchor_ori_b = R_to_rot6d(quat_to_R(ori_q))
        return motion_anchor_ori_b[0].astype(np.float32)
```

https://github.com/user-attachments/assets/8f9af28f-a88a-4d6a-87b4-491935b3c66f

It works!
